plot/vg/draw: introduce and use TextHandler

sbinet opened the pull request:
Please take a look.

<!--
Checklist:

- API changes have been discussed
- code is goformated correctly (goimports)
- packages with generated code have had code generation run
- tests pass locally
- linked to relevant issues

Please make sure your commit message summary line and pull request
title match the Go convention; a one-line summary of the change,
prefixed by the primary affected package that should complete the
sentence, "This change modifies Gonum to _____."
-->

sbinet commented:
ping?

kortschak commented:
This will take some time to look through. I will try to get to it in the next couple of days (task added to personal assistant).

sbinet commented:
I've dropped the HTML handler out of this PR (to reduce the number of files to review)
If we proceed with this PR (or some version of it), I'll file it back in another one.

(I've also rebased off the latest master, and removed the `panic`s for the `Canvas.DPI` implementations (putting a default/hardcoded 72 value) for canvas implementations that do not expose/use an explicit DPI value)

kortschak commented on gob/gob.go:
```diff
@@ -43,4 +44,7 @@ func init() {
 	gob.Register(plotter.XYValues{})
 
+	// vg/draw.TextStyle
+	gob.Register(plot.DefaultTextHandler)
+	gob.Register(text.Plain{})
```
`plot.DefaultTextHandler == text.Plain{}`; why do we register both?

sbinet replied:
mainly for "forward coding" (in the perhaps remote case when we'd switch DefaultTextHandler to default to another implementation)

I can remove if you prefer.

kortschak replied:
OK. That's fine.

kortschak commented on text/latex.go:
```diff
@@ -0,0 +1,143 @@
+	"github.com/go-latex/latex/font/ttf"
+	"github.com/go-latex/latex/mtex"
+	"github.com/go-latex/latex/tex"
+	"gonum.org/v1/plot/vg"
```
Insert blank line.

kortschak commented on text/latex.go:
```diff
@@ -0,0 +1,143 @@
+}
+
+func (r *latex) Render(width, height, dpi float64, c *drawtex.Canvas) error {
+	//var (
+	//	w = width * dpi
+	//	h = height * dpi
+	//)
+
```
Delete?

kortschak commented on text/latex.go:
```diff
@@ -0,0 +1,143 @@
+	r.cnv.FillPolygon(r.sty.Color, pts)
+}
+
+var (
+	_ draw.TextHandler = (*Latex)(nil)
+	_ mtex.Renderer    = (*latex)(nil)
+)
```
Can you move this to the top? It should be a documentation of the interface that the type is intended to satisfy.

kortschak commented on text/plain.go:
```diff
@@ -0,0 +1,16 @@
+
+package text // import "gonum.org/v1/plot/text"
+
+import (
+	"gonum.org/v1/plot/vg/draw"
+)
```
`import "gonum.org/v1/plot/vg/draw"`

kortschak commented on text/plain.go:
```diff
@@ -0,0 +1,16 @@
+// Plain is a text/plain handler.
+type Plain = draw.PlainTextHandler
+
+var (
+	_ draw.TextHandler = (*Plain)(nil)
+)
```
`var _ draw.TextHandler = (*Plain)(nil)`

kortschak commented on text/text.go:
```diff
@@ -0,0 +1,6 @@
+// Copyright ©2020 The Gonum Authors. All rights reserved.
```
I think usually we call these "doc.go".

kortschak commented on vg/recorder/recorder.go:
```diff
@@ -112,6 +112,11 @@ func (c *Canvas) append(a Action) {
 }
 
+func (c *Canvas) DPI() float64 {
+	// FIXME(sbinet): add a way to specify a value?
```
If we need to add these we can either make canvases require a constructor which takes a DPI, or add a resolution field that is interpreted as 72 if it's 0, or make it a conditional interface where 72 is the default value if there is no DPI method. Given that LaTeX is the only one that needs it, the last approach feels the nicest to me.

sbinet replied:
I am a bit worried a `xyz.Canvas` implementing this `DPI()` method could be wrapped by a couple layers of `vg.Canvas` (sans the `DPI()` method) interfaces and then presented to a `draw.Canvas` value.
I think the type switch wouldn't be able to dig that `DPI()` method out of the interface layers.

but perhaps it's just a remote, theoretical and uncommon failure mode....

kortschak commented on vg/vg.go:
```diff
@@ -17,6 +17,8 @@ import (
 // graphics.
 // The origin is in the bottom left corner.
 type Canvas interface {
+	DPI() float64
+
```
Why did we need to add the `DPI` method?

sbinet replied:
for the text.Latex handler.

sbinet commented:
ping?

kortschak commented:
I'm still concerned about the `DPI` method.

sbinet commented:
PTAL: removed DPI from vg.Canvas, set as a field of Latex.

kortschak commented on text/latex.go:
```diff
@@ -18,15 +18,17 @@ import (
 // Latex parses, formats and renders LaTeX.
-type Latex struct{}
+type Latex struct {
+	DPI float64
```
Add a doc comment that this sets that LaTeX font resolution and also that resolution defaults to 72 if this is unset.

sbinet replied:
done
